src/storage/host_store.rs

```diff
@@ -101,13 +111,11 @@ impl PageStorage {
     }
 
     fn pages_data_len(&self) -> usize {
-        match self.pages.len() {
-            0 => 0,
-            n => {
-                (n - 1) * PAGE_SIZE
-                    + self.pages.last().map(|p| p.slice().len()).unwrap_or(0)
-            }
+        let mut size_sum = 0;
+        for p in &self.pages {
```

Comment:
You're changing from a constant time to a linear time function here, this should be avoided